### PR TITLE
Add TCP support

### LINKED ISSUE #2: Add TCP support

### PR DESCRIPTION
**Pull request by ammarfaizi2**
Add TCP support.

This PR will resolve #2.

### CODE REVIEW

**Comment by ammarfaizi2** [comment]
@zeheater @louvian 
I need test and review for this PR. It works on my machine but is not fast as OpenVPN.

##  Setup Instructions
```
# Clone the branch dev_001.
git clone -b dev_001 https://github.com/TeaInside/teavpn2;
cd teavpn2;

# Build.
make -j$(nproc);

# Copy config template
cp -v config/client.ini client.tmp

# Edit the username (available usernames: admin, ammarfaizi2, louvian, zeheater, soewako).
sed -i 's/ammarfaizi2/username/' client.tmp

# Run teavpn client (need root permission to open /dev/net/tun).
sudo ./teavpn_client -c client.tmp;
```

## Output
```
ammarfaizi2@integral:~/project/now/teavpn2$ sudo ./teavpn_client -c client.tmp
Parsing arguments...
  config->config_file = "client.tmp"
[Mon Oct  5 20:00:36 2020]: Validating client config...
[Mon Oct  5 20:00:36 2020]: Allocating virtual network interface...
[Mon Oct  5 20:00:36 2020]: Allocating tun_fd...
[Mon Oct  5 20:00:36 2020]: Initializing pipe...
[Mon Oct  5 20:00:36 2020]: Initializing TCP socket...
[Mon Oct  5 20:00:36 2020]: Creating TCP socket...
[Mon Oct  5 20:00:36 2020]: TCP socket created successfully!
[Mon Oct  5 20:00:36 2020]: Setting up socket file descriptor...
[Mon Oct  5 20:00:36 2020]: Socket file descriptor set up successfully!
[Mon Oct  5 20:00:36 2020]: Connecting to 68.183.184.174:55555...
[Mon Oct  5 20:00:36 2020]: Connection established!
[Mon Oct  5 20:00:36 2020]: Authenticating...
[Mon Oct  5 20:00:36 2020]: recv(): 14 bytes from net_fd.
[Mon Oct  5 20:00:36 2020]: Got SRV_PKT_AUTH_OK!
[Mon Oct  5 20:00:36 2020]: Authentication success!
[Mon Oct  5 20:00:36 2020]: Executing: /sbin/ip link set dev 'teavpn00' up mtu 1500
[Mon Oct  5 20:00:36 2020]: Executing: /sbin/ip addr add dev 'teavpn00' '10.10.10.3/24' broadcast '10.10.10.255'
[Mon Oct  5 20:00:36 2020]: recv(): 86 bytes from net_fd.
[Mon Oct  5 20:00:36 2020]: Write to tun_fd 80 bytes
[Mon Oct  5 20:00:36 2020]: Read from tun_fd 52 bytes
[Mon Oct  5 20:00:36 2020]: Send to net_fd 58 bytes
[Mon Oct  5 20:00:36 2020]: recv(): 86 bytes from net_fd.
[Mon Oct  5 20:00:36 2020]: Write to tun_fd 80 bytes
...
...
...
```
## Test Ping to Server over VPN
```
ammarfaizi2@integral:~/project/now/teavpn2$ ping 10.10.10.1
PING 10.10.10.1 (10.10.10.1) 56(84) bytes of data.
64 bytes from 10.10.10.1: icmp_seq=1 ttl=64 time=277 ms
64 bytes from 10.10.10.1: icmp_seq=2 ttl=64 time=127 ms
64 bytes from 10.10.10.1: icmp_seq=3 ttl=64 time=139 ms
64 bytes from 10.10.10.1: icmp_seq=4 ttl=64 time=157 ms
64 bytes from 10.10.10.1: icmp_seq=5 ttl=64 time=195 ms
^C
--- 10.10.10.1 ping statistics ---
6 packets transmitted, 5 received, 16,6667% packet loss, time 5007ms
rtt min/avg/max/mdev = 126.903/179.004/277.334/54.239 ms
ammarfaizi2@integral:~/project/now/teavpn2$ 

```

## Test Internet Connection Forward
```
sudo ip route add 0.0.0.0/1 via 10.10.10.1 dev teavpn00;
sudo ip route add 128.0.0.0/1 via 10.10.10.1 dev teavpn00;
sudo ip route add 10.10.10.0/24 via 10.10.10.1 dev teavpn00;

DEFAULT_GW=$(ip route | grep "default via" | awk '{print $3}');
sudo ip route add 68.183.184.174/32 via $DEFAULT_GW;

# Check your IP.
curl ipecho.net/plain;echo;

# You can do other things that need internet connectivity.
```

## iperf test
I am also serving iperf on port 22.
```
root@integral:~# iperf -c 10.10.10.1 -p 22
------------------------------------------------------------
Client connecting to 10.10.10.1, TCP port 22
TCP window size: 85.0 KByte (default)
------------------------------------------------------------
[  3] local 10.10.10.3 port 36730 connected with 10.10.10.1 port 22
[ ID] Interval       Transfer     Bandwidth
[  3]  0.0-10.0 sec  1.00 MBytes   838 Kbits/sec
root@integral:~# 
```

File: src/include/teavpn2/global/common.h [hunk 1]
```diff
@@ -5,12 +5,82 @@
...
+int tun_alloc(char *dev, int flags);
+int tun_set_queue(int fd, bool enable);
+uint8_t cidr_jump_table(__be32 netmask);
+int fd_set_nonblock(int fd);
+
+#define TCP_RECV_BUFFER DATA_SIZE
+#define TCP_SEND_BUFFER DATA_SIZE
+
+#define TCP_RECV_SIZE (DATA_SIZE)
+#define TCP_SEND_SIZE (DATA_SIZE)
```
Comment (louvian) [inline src/include/teavpn2/global/common.h:84]: Useless meaning of constants represent the same value.

File: src/sources/teavpn2/client/sock/tcp/client.c [hunk 2]
```diff
@@ -0,0 +1,568 @@
...
+       * Not only zero the recv_size, the callee is also
+       * responsible to zero the buffer, since it contains
+       * the length of data. This length of data may be reused
+       * if it is not zeroed.
+       */
+
+      case SRV_PKT_PING:
+        srv_pkt->size    = 0;
+        state->recv_size = 0;
+        break;
```
Comment (louvian) [inline src/sources/teavpn2/client/sock/tcp/client.c:397]: Missing event handler.

File: src/include/teavpn2/global/common.h [hunk 3]
```diff
@@ -5,12 +5,82 @@
...
+  srv_packet_type   type;
+  uint16_t          size;                     /* The size of data to be sent. */
+  char              data[DATA_SIZE];
+} server_pkt;
+
+#define CLI_IDENT_PKT_SIZE (OFFSETOF(client_pkt, data))
+#define SRV_IDENT_PKT_SIZE (OFFSETOF(client_pkt, data))
+
+#define CL_IDENT_SIZ (OFFSETOF(client_pkt, data))
+#define SR_IDENT_SIZ (OFFSETOF(server_pkt, data))
```
Comment (louvian) [inline src/include/teavpn2/global/common.h:71]: Useless meaning of constants represent the same value.

File: src/include/teavpn2/global/common.h [hunk 4]
```diff
@@ -5,12 +5,82 @@
...
+  cli_packet_type   type;
+  uint16_t          size;                     /* The size of data to be sent. */
+  char              data[DATA_SIZE];
+} client_pkt;
+
+typedef struct __attribute__((__packed__)) _server_pkt {
+  srv_packet_type   type;
+  uint16_t          size;                     /* The size of data to be sent. */
+  char              data[DATA_SIZE];
+} server_pkt;
```
Comment (louvian) [inline src/include/teavpn2/global/common.h:65]: To keep it SOLID, please separate these struct definitions into several files.

**Review by louvian — CHANGES_REQUESTED** [review]
Code review.

**Comment by louvian** [comment]
Hi @ammarfaizi2,
I ran it and it worked well, although the transfer is not good as a general famous VPN software.

I don't know how much of the earlier discussion you did at the time, or have looked back at, but it's worth that the TCP version passed iperf test with a better result after several long commits you have done.

The lack of multi-client support wasn't an accident, but it needs more research and tests. It was part of how approval was achieved. However, that does not mean I can't approve this PR.

Rather than "missing the point", I think it's a case of different main usage of this software. I suggested three perspectives for us to focus on:
1. Stability of TCP connection.
2. Encryption algorithm implementation (it should also include compression to reduce overhead I think).
3. Multi-client support.

You did a great work until it the TCP connection support is back to TeaVPN.

Regards,
Louvian Lyndal.

**Comment by ammarfaizi2** [comment]
Thanks for the comments and reviews. I will address the requested changes in the near future.
I have a lot of work these days, can't maintaining it.

**Comment by ammarfaizi2** [comment]
This PR should also resolve #14

**Comment by louvian** [comment]
The CI doesn't run because of git conflict. Can we merge these changes first, then resolve the conflicts, then re-create a new PR @ammarfaizi2?

**Comment by ammarfaizi2** [comment]
Yeah, I think it is ok to merge this PR to `master` branch even if the current version is not working yet. No one will get affected.

**Comment by ammarfaizi2** [comment]
I will merge it.

**Pull request merged**